### DISCUSSION TITLE
missing Response.ok method

### DISCUSSION THREAD

**Discussion by codenoid** [encode/httpx#1521] [General]
`resp.ok` is cute

**Comment by lovelydinosaur** [comment]
We've got a comment in the "requests compatibility" docs about this.

**Comment by StephenBrown2** [comment]
Link to aforementioned docs: https://www.python-httpx.org/compatibility#checking-for-4xx5xx-responses

**Comment by serg06** [comment]
`if not resp.is_error` just doesn't hit the same as `if resp.ok` 🙁

**Comment by Colin-b** [comment]
See https://github.com/encode/httpx/issues/809

I fail to see a use case were you would want to discard any errors (no logs, no treatment at all) and only perform action in case of successful response. This would be the only use case for a `if response.ok` and it seems like bad practice. Do you have such use case ?

**Comment by serg06** [comment]
> See #809
> 
> I fail to see a use case were you would want to discard any errors (no logs, no treatment at all) and only perform action in case of successful response. This would be the only use case for a `if response.ok` and it seems like bad practice. Do you have such use case ?

I don't, I'm just a sassy opinionated coder

**Comment by Colin-b** [comment]
> > See #809
> > I fail to see a use case were you would want to discard any errors (no logs, no treatment at all) and only perform action in case of successful response. This would be the only use case for a `if response.ok` and it seems like bad practice. Do you have such use case ?
> 
> I don't, I'm just a sassy opinionated coder

So you don't need to use `if not`, you can replace your:

```python
import httpx

response = httpx.get()
if not response.is_error:
    handle_success()
else:
    handle_error()
```

by:

```python
import httpx

response = httpx.get()
if response.is_error:
    handle_error()
else:
    handle_success()
```

And then you have it, no negation involved :)

**Comment by serg06** [comment]
> And then you have it, no negation involved :)

But then I'd be putting the (unlikely) error case before the (likely) success case!

**Comment by lovelydinosaur** [comment]
> `if not resp.is_error` just doesn't hit the same as `if resp.ok` 🙁 

Them's the breaks.

**Comment by lovelydinosaur** [comment]
So, the reason we don't do this is pretty clear.
As a developer, what do you think `response.ok` *means*?

* `response.status_code == 200` *A "200 OK" response*
* `response.status_code >= 200 and response.status_code <= 299` *Any 2xx response*
* `not response.is_error` *Not a "4xx, 5xx" response*

It's just not at all clear. (Incidentally, in requests it means the last one of those.)

I could *perhaps* see us having `response.is_success`, `response.is_client_error` and `response.is_server_error` properties., which are *way* more obviously 2xx, 4xx, and 5xx classes of response. But... note that would open up easy coding error classes like...

```python
response = httpx.get("https://www.example.com", allow_redirects=False)
if response.is_success:
    ...  # Success condition
else:
    ...  # Error condition
```

Which doesn't do what it might look like it does on first sight.

So... sure `if not response.is_error` ain't got as much sass as `if response.ok`, or even `if response.is_success`, but it'll help prevent coding errors.

(I mean there's other options too. For instance, `response.is_2xx`, `response.is_3xx`, etc... but in terms of requests-parity I'm pretty convinced we *don't* want the misleading `response.ok`)
